builtin/credential/aws/path_login.go

```diff
@@ -888,6 +1030,392 @@ func (b *backend) pathLoginRenew(
+	return false, "Missing Authorization header"
+}
+
+func buildHttpRequest(method, endpoint string, parsedUrl *url.URL, body string, headers http.Header) *http.Request {
+	// This is all a bit complicated because the AWS signature algorithm requires that
+	// the Host header be included in the signed headers. See
+	// http://docs.aws.amazon.com/general/latest/gr/sigv4-create-canonical-request.html
+	// The use cases we want to support, in order of increasing complexity, are:
+	// 1. All defaults (client assumes sts.amazonaws.com and server has no override)
+	// 2. Alternate STS regions: client wants to go to a specific region, in which case
+	//    Vault must be confiugred with that endpoint as well. The client's signed request
+	//    will include a signature over what the client expects the Host header to be,
+	//    so we cannot change that and must match.
+	// 3. Alternate STS regions with a proxy that is transparent to Vault's clients.
+	//    In this case, Vault is aware of the proxy, as the proxy is configured as the
+	//    endpoint, but the clients should NOT be aware of the proxy (because STS will
+	//    not be aware of the proxy)
+	// It's also annoying because:
+	// 1. The AWS Sigv4 algorithm requires the Host header to be defined
+	// 2. Some of the official SDKs (at least botocore and aws-sdk-go) don't actually
+	//    incude an explicit Host header in the HTTP requests they generate, relying on
+	//    the underlying HTTP library to do that for them.
+	// 3. To get a validly signed request, the SDKs check if a Host header has been set
+	//    and, if not, add an inferred host header (based on the URI) to the internal
+	//    data structure used for calculating the signature, but never actually expose
+	//    that to clients. So then they just "hope" that the underlying library actually
+	//    adds the right Host header which was included in the signature calculation.
+	// We could either explicity require all Vault clients to explicitly add the Host header
+	// in the encoded request, or we could also implicitly infer it from the URI.
+	// We choose to support both -- allow you to explicitly set a Host header, but if not,
+	// infer one from the URI.
+	// HOWEVER, we have to preserve the request URI portion of the client's
+	// URL because the GetCallerIdentity Action can be encoded in either the body
+	// or the URL. So, we need to rebuild the URL sent to the http library to have the
+	// custom, Vault-specified endpoint with the client-side request parameters.
+	targetUrl := fmt.Sprintf("%s/%s", endpoint, parsedUrl.RequestURI())
+	request, err := http.NewRequest(method, targetUrl, strings.NewReader(body))
+	if err != nil {
+		return nil
+	}
+	request.Host = parsedUrl.Host
+	for k, vals := range headers {
+		for _, val := range vals {
+			request.Header.Add(k, val)
+		}
+	}
+	return request
+}
+
+func ensureHeaderIsSigned(signedHeaders, headerToSign string) (bool, string) {
```

Comment:
Similar to comments on other functions, this can also return just the `error` type.